File: src/nappe/diff.py
from __future__ import annotations

import json
from typing import Sequence

from nappe.rules import FixSafety, Suggestion


def format_text(suggestions: Sequence[Suggestion]) -> str:
    if not suggestions:
        return ""
    lines: list[str] = []
    for s in suggestions:
        lines.append(
            f"{s.file_path}:{s.line}:{s.col}: {s.rule.code} {s.rule.description}"
        )
        if s.context:
            lines.append(f"    {s.line} | {s.context}")
            marker = " " * (len(str(s.line)) + 4) + "^" * len(s.context.strip())
            lines.append(f"{marker} {s.rule.code}")
        lines.append("")
    safe = sum(1 for s in suggestions if s.rule.safety == FixSafety.SAFE)
    unsafe = sum(1 for s in suggestions if s.rule.safety == FixSafety.UNSAFE)
    lines.append(
        f"Found {len(suggestions)} issue{'s' if len(suggestions) != 1 else ''} ({safe} safe, {unsafe} unsafe)."
    )
    return "\n".join(lines)


def format_diff(suggestions: Sequence[Suggestion]) -> str:
    if not suggestions:
        return ""
    lines: list[str] = []
    for s in suggestions:
        lines.append(f"--- a/{s.file_path}")
        lines.append(f"+++ b/{s.file_path}")
        lines.append(f"@@ -{s.line},1 +{s.line},1 @@")
        old_line = s.context
        new_lines = s.new_source.decode("utf-8", errors="replace").split("\n")
        new_line = new_lines[0] if new_lines else ""
        lines.append(f"-{old_line}")
        lines.append(f"+{new_line}")
        lines.append("")
    return "\n".join(lines)


def format_json(suggestions: Sequence[Suggestion]) -> str:
    data = []
    for s in suggestions:
        data.append(
            {
                "file": s.file_path,
                "line": s.line,
                "col": s.col,
                "code": s.rule.code,
                "message": s.rule.description,
                "safety": s.rule.safety.value,
            }
        )
    return json.dumps(data, indent=2)


def apply_fixes(
    suggestions: Sequence[Suggestion], safety: FixSafety
) -> dict[str, bytes]:
    by_file: dict[str, list[Suggestion]] = {}
    for s in suggestions:
        if s.rule.safety == safety:
            by_file.setdefault(s.file_path, []).append(s)
    results: dict[str, bytes] = {}
    for file_path, file_suggestions in by_file.items():
        source = file_suggestions[0].old_source
        for s in sorted(file_suggestions, key=lambda x: x.line, reverse=True):
            source = s.new_source
        results[file_path] = source
    return results

File: src/diff.rs
use crate::rules::{FixSafety, Suggestion};

pub fn format_text(suggestions: &[Suggestion]) -> String {
    if suggestions.is_empty() {
        return String::new();
    }
    let mut lines = Vec::new();
    for s in suggestions {
        lines.push(format!(
            "{}:{}:{}: {} {}",
            s.file_path, s.line, s.col, s.rule.code, s.rule.description
        ));
        if !s.context.is_empty() {
            lines.push(format!("    {} | {}", s.line, s.context));
            let marker_len = format!("{}", s.line).len() + 4;
            let trimmed = s.context.trim();
            let marker = " ".repeat(marker_len) + &"^".repeat(trimmed.len());
            lines.push(format!("{} {}", marker, s.rule.code));
        }
        lines.push(String::new());
    }
    let safe = suggestions
        .iter()
        .filter(|s| s.rule.safety == FixSafety::Safe)
        .count();
    let unsafe_count = suggestions
        .iter()
        .filter(|s| s.rule.safety == FixSafety::Unsafe)
        .count();
    lines.push(format!(
        "Found {} issue{} ({} safe, {} unsafe).",
        suggestions.len(),
        if suggestions.len() != 1 { "s" } else { "" },
        safe,
        unsafe_count
    ));
    lines.join("\n")
}

pub fn format_diff(suggestions: &[Suggestion]) -> String {
    if suggestions.is_empty() {
        return String::new();
    }
    let mut lines = Vec::new();
    for s in suggestions {
        lines.push(format!("--- a/{}", s.file_path));
        lines.push(format!("+++ b/{}", s.file_path));
        lines.push(format!("@@ -{},1 +{},1 @@", s.line, s.line));
        lines.push(format!("-{}", s.context));
        let new_lines: Vec<&str> = std::str::from_utf8(&s.new_source)
            .unwrap_or("")
            .split('\n')
            .collect();
        let new_line = new_lines.first().copied().unwrap_or("");
        lines.push(format!("+{}", new_line));
        lines.push(String::new());
    }
    lines.join("\n")
}

pub fn format_json(suggestions: &[Suggestion]) -> String {
    let items: Vec<String> = suggestions
        .iter()
        .map(|s| {
            format!(
                r#"  {{ "file": "{}", "line": {}, "col": {}, "code": "{}", "message": "{}", "safety": "{}" }}"#,
                s.file_path, s.line, s.col, s.rule.code, s.rule.description, s.rule.safety.as_str()
            )
        })
        .collect();
    format!("[\n{}\n]", items.join(",\n"))
}

pub fn apply_fixes(suggestions: &[Suggestion], safety: FixSafety) -> Vec<(String, Vec<u8>)> {
    let mut by_file: Vec<(&str, &Suggestion)> = suggestions
        .iter()
        .filter(|s| s.rule.safety == safety)
        .map(|s| (s.file_path.as_str(), s))
        .collect();
    by_file.sort_by_key(|(_, s)| s.file_path.clone());
    by_file.dedup_by_key(|(_, s)| s.file_path.clone());

    let mut results = Vec::new();
    for (file_path, _) in &by_file {
        let file_suggestions: Vec<&Suggestion> = suggestions
            .iter()
            .filter(|s| s.file_path == *file_path && s.rule.safety == safety)
            .collect();
        if let Some(first) = file_suggestions.first() {
            let mut source = first.old_source.clone();
            for s in file_suggestions.iter().rev() {
                source = s.new_source.clone();
            }
            results.push((file_path.to_string(), source));
        }
    }
    results
}
